examples-next/auth/schema.ts

```diff
@@ -0,0 +1,144 @@
+import { createSchema, list } from '@keystone-spike/keystone/schema';
+import { text, checkbox, password } from '@keystone-spike/fields';
+
+/**
+ * TODO: with the new session API, we're breaking keystone's current assumptions about the
+ * `authentication` argument provided to access control functions. This will be a breaking change
+ * for developers using Keystone.
+ *
+ * This needs to be implemented (replace `authentication` with `session`) before the code below
+ * will work.
+ *
+ * While we're here I'd like to propose some other changes to improve the ergonomics; see the types
+ * below for my proposal. Note that I'm specifically omitting the graphql context and any way of
+ * executing queries because I'd like to review that API as well, separately (and we can add it in
+ * later). This may be a bit contentious, but we need a new object anyway so I'm happy if in the
+ * first pass we just implement ({ session, item }) argument so the code below works and we can
+ * discuss the rest of the proposal as a group.
+ *
+ * Note that custom schema access control deviates significantly from the list and field access
+ * control arguments; we need to revisit that in the context of how graphql schema extensions work
+ * now. See https://www.keystonejs.com/api/access-control#custom-schema-access-control
+ *
+ * Also note, I'm not sure the reasoning behind why ListAccess seems to be resolved once per
+ * mutation for mutations that operate on multiple items, while FieldAccess seems to be resolved
+ * once per item (many times per mutation) for the same. At least, that's what I think is going on.
+ *
+ * This seems like a real foot-gun, and I'd prefer that they're implemented consistently, in favor
+ * of aligning on FieldAccessInput and resolving the function once per item being updated in the
+ * case of mutations that operate on multiple items.
+ *
+ * Or maybe my understanding is wrong? This is what's implied by the documentation (see
+ * https://www.keystonejs.com/api/access-control#field-level-access-control) but then if it's not
+ * the case I don't see how `existingItem` works in mutations that operate on multiple items.
+ *
+ * I also have questions about `originalInput` in the case of nested mutations (e.g creating an
+ * item as the input to a relationship field). Is it the whole "original" input? Or just the nested
+ * input for the relationship? Seems like it would have to be the second, but it's not clear from
+ * the documentation. Is this consistent with how the `gqlName` argument works?
+ *
+ * tbh I'm inclined to remove the gqlName argument unless there's a compelling reason to have it;
+ * esp. given it's marked as optional.
```

Comment:
`gqlName` argument is useful to know the difference between calling `allUsers` vs `_allUsersMeta`, for example.

Reply:
Does that actually deliver value? You'd need to (I imagine) handle _every_ case for the top level gqlName if it was important in any way... ie you might resolve permissions for users through a relationship to content, in which case the top-level query name might be `allPosts` 🤷‍♂️ 

If distinguishing between querying items and meta is important, we could pass that in, but I also don't _really_ get the use case?

Reply:
Or maybe I've misinterpreted the docs:

> The name of the query or mutation which triggered the access check.

But still, then in the following query don't I need to account for the fields that trigger the query?

e.g if I'm writing an access check on users, and the query looks like:

```
query {
  allPosts {
    author {
      name
    }
  }
}
```

Is the gqlName `allPosts` / `author` or something else?

or in this case:

```
query {
  allUsers {
    _postsMeta {
      count
    }
  }
}
```

What's the `gqlName` for the `Post` list access check? `allUsers` / `_postsMeta` etc? in either case this seems super hard because you're encoding knowledge of all the relationships into access control.